keps/238-controller-revision/README.md

```diff
@@ -0,0 +1,393 @@
+spec when restarted during rolling update, and be able to compare different LWS objects to determine if the spec has changed. 
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+#### Story 2
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+There are a few constraints that need to be taken into account in the design:
+- The label `templateHash` is not a reliable way to determine whether or not an LWS object has been updated, as seen in #281
+- The same LWS object can generate two different controllerRevision names
+	- For instance, everytime the LWS is queried, it will have different `lws.ResourceVersion`, causing two revisions to be named differently even if they have the same LWS object
+- Adding a new default pod label will trigger rolling update when updating from a different LWS controller version, so no new pod labels can be added.
+
+Taking those into account, we'll combine controller revisions and template hashes. Each controller revision will have a templateHash as a label, essentially creating a map of revisions with template hashes as keys. 
+
+
+### LWS controller
+In order to fix #281, we will use the hash key in the leaderSts to make a controller revision lookup, and get the LWS object that was used to create it. To determine if an update has happened, we'll compare the fields that are used to generate the template hash.
+
+
+```golang
+func templateUpdated(sts, lws) bool {
+	controllerRevision := GetLeaderWorkerSetRevisionFromTemplateHash(sts.Labels[templateHash])
+	baselineLws:= controllerutils.ApplyRevision(lws, controllerRevision)
+	return !utils.EqualLeaderWorkerTemplates(baselineLws, lws)
```

Comment:
we also need to check some parameters of the LWS spec itself, specifically the subdomain policy

Reply:
Yes, `equalLeaderWorkerTemplates` would check both `lws.spec.leaderWorkerTemplate` and checking `subdomainPolicy`, also treating `networkConfig == nil` as `subdomainShared` just like it is done for `templateHash`

Reply:
Yes, equalLeaderWorkerTemplates would check both lws.spec.leaderWorkerTemplate and checking subdomainPolicy, also treating networkConfig == nil as subdomainShared just like it is done for templateHash